wallet-next/src/wallet.rs

```diff
@@ -0,0 +1,144 @@
+use anyhow::Context;
+use penumbra_crypto::{
+    fmd,
+    keys::{
+        FullViewingKey, IncomingViewingKey, OutgoingViewingKey, SeedPhrase, SpendKey, SpendSeed,
+    },
+    Address, Note,
+};
+use serde::{Deserialize, Serialize};
+
+/// The contents of the wallet file that share a spend authority.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(from = "serde_helpers::WalletHelper")]
+#[serde(into = "serde_helpers::WalletHelper")]
+pub struct Wallet {
+    /// A list of human-readable labels for addresses.
+    ///
+    /// The label at index `i` is used for the address with `DiversifierIndex(i)`.
+    address_labels: Vec<String>,
+    spend_key: SpendKey,
+}
+
+impl Wallet {
+    /// Create a new wallet.
+    pub fn from_seed_phrase(seed_phrase: SeedPhrase) -> Self {
+        // Currently we support a single spend authority per wallet. In the future,
+        // we can derive multiple spend seeds from a single seed phrase.
+        let spend_seed = SpendSeed::from_seed_phrase(seed_phrase, 0);
+        let spend_key = SpendKey::new(spend_seed);
+
+        Self {
+            spend_key,
+            address_labels: vec!["Default".to_string()],
+        }
+    }
+
+    /// Imports a wallet from a legacy [`SpendSeed`].
+    pub fn import(spend_seed: SpendSeed) -> Self {
+        let spend_key = spend_seed.into();
+        Self {
+            spend_key,
+            address_labels: vec!["Default".to_string()],
+        }
+    }
+
+    /// Incoming viewing key from this spend seed.
+    pub fn incoming_viewing_key(&self) -> &IncomingViewingKey {
+        self.spend_key.full_viewing_key().incoming()
+    }
+
+    /// Outgoing viewing key from this spend seed.
+    pub fn outgoing_viewing_key(&self) -> &OutgoingViewingKey {
+        self.spend_key.full_viewing_key().outgoing()
+    }
+
+    /// Returns the wallet's spend seed.
+    pub fn spend_key(&self) -> &SpendKey {
+        &self.spend_key
+    }
+
+    /// Get the full viewing key for this wallet.
+    pub fn full_viewing_key(&self) -> &FullViewingKey {
+        self.spend_key.full_viewing_key()
+    }
+
+    /// Generate a new diversified `Address` and its corresponding `DetectionKey`.
+    pub fn new_address(&mut self, label: String) -> (usize, Address, fmd::DetectionKey) {
+        let next_index = self.address_labels.len();
+        self.address_labels.push(label);
+        let (address, dtk) = self
+            .incoming_viewing_key()
+            .payment_address(next_index.into());
+        (next_index, address, dtk)
+    }
+
+    /// Get address by index.
+    pub fn address_by_index(&self, index: usize) -> Result<(String, Address), anyhow::Error> {
+        let label = self
+            .address_labels
+            .get(index)
+            .ok_or_else(|| anyhow::anyhow!("no address with index {}", index))?;
```

Comment:
The address labels are just client-side information, so if we kept them in the wallet service, we'd need a way to sync them back and forth between the wallet client and the wallet service.

But there's no need for the wallet service to track them at all.  Instead of working with the labels, or even with the indexes, the wallet service should just work with 11-byte [`DiversifierIndex`es](https://rustdoc.penumbra.zone/main/penumbra_crypto/keys/struct.DiversifierIndex.html), and let the wallet client figure out how to interpret them.